library/alloc/src/string.rs

```diff
@@ -2272,11 +2272,7 @@ impl<T: fmt::Display + ?Sized> ToString for T {
     // to try to remove it.
     #[inline]
     default fn to_string(&self) -> String {
-        use fmt::Write;
-        let mut buf = String::new();
-        buf.write_fmt(format_args!("{}", self))
-            .expect("a Display implementation returned an error unexpectedly");
-        buf
+        format!("{}", self)
```

Comment:
`format!()` uses `String::with_capacity(_)` internally, not sure if that is a better behavior here.

Reply:
`write_fmt` calls `write_str` multiple times, which is potentially multiple allocations here as no capacity is allocated up front. It seems to me that `with_capacity` would be better in all cases? Or is there a case where the current behavior is more efficient?

Reply:
`format!()` calls `write_fmt` internally or maybe I was looking at stable rust code?

Reply:
`format!()` calls `write_fmt` internally, but allocates capacity up front, as opposed to the current behavior in `to_string` where no capacity is allocated up front.

Reply:
Why not compare asm at least?

Reply:
@ibraheemdev `format!` [gets](https://doc.rust-lang.org/src/alloc/fmt.rs.html#575-579) the initial capacity from `fn estimated_capacity()` on `Arguments`, which is very [simple heuristic](https://doc.rust-lang.org/src/core/fmt/mod.rs.html#352-355) that triggers when the format string **mixes string literals and placeholders** (eg. `format!("hello, {}", x)` – in such case we can assume that the final string will most likely be strictly longer than then literal part and thus if `x` is short enough, avoid reallocation).

Anyway, the current implementation will simply return 0 initial capacity for `format!({})` (even if the `fmt` implementation does something like `write!(f, "hello, {}", x)`) – the capacity estimation is not recursive. So, with current capacity estimation heuristics, there won't be any difference when calling `format!`.

Reply:
@krdln Ah, I see. I still think this PR should be be merged in case the capacity estimation is updated in the future.